model/test_results.go

```diff
@@ -173,50 +171,43 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
 		true,
 	)
 	if err != nil {
-		return errors.Wrap(err, "problem creating bucket")
-	}
-
-	resultChan := make(chan *TestResult, len(results))
-	for i := 0; i < len(results); i++ {
-		resultChan <- &results[i]
-	}
-	close(resultChan)
-	var wg sync.WaitGroup
-	catcher := grip.NewBasicCatcher()
-	for i := 0; i < runtime.NumCPU(); i++ {
-		wg.Add(1)
-		go func() {
-			defer func() {
-				catcher.Add(recovery.HandlePanicWithError(recover(), nil, "test results append"))
-				wg.Done()
-			}()
-
-			for result := range resultChan {
-				data, err := bson.Marshal(result)
-				if err != nil {
-					// We should continue on marhsalling
-					// errors so we can get as many test
-					// results in s3 as possible.
-					catcher.Wrapf(err, "problem marshalling bson for test result '%s'", result.TestName)
-					continue
-				}
-
-				if err := bucket.Put(ctx, result.TestName, bytes.NewReader(data)); err != nil {
-					catcher.Wrapf(err, "problem uploading test result '%s' to bucket", result.TestName)
-					return
-				}
-			}
-		}()
-	}
-	wg.Wait()
-
-	return catcher.Resolve()
+		return errors.Wrap(err, "creating bucket")
+	}
+
+	var allResults testResultsDoc
+	r, err := bucket.Get(ctx, testResultsCollection)
+	if err != nil && !pail.IsKeyNotFoundError(err) {
+		return errors.Wrap(err, "getting uploaded test results")
+	}
+	if r != nil {
+		defer r.Close()
+
+		data, err := ioutil.ReadAll(r)
+		if err != nil {
+			return errors.Wrap(err, "reading uploaded test results")
+		}
+
+		if err = bson.Unmarshal(data, &allResults); err != nil {
+			return errors.Wrap(err, "unmarshalling uploaded test results")
+		}
+	}
+	allResults.Results = append(allResults.Results, results...)
+
+	data, err := bson.Marshal(&allResults)
```

Comment:
How large could the file get? If it's really big, would we consider using [gob](https://golang.org/pkg/encoding/gob/) to encode the results? Some [people on the internet](https://stackoverflow.com/a/41180680) seem to think it's more efficient.

Reply:
I have this same question. The serialize/update/deserialize roundtrip seems okay but only if it won't take up a ton of memory for large test results.

Reply:
Given the size of a single test result is < 500B, even if a task has 10k or even 20k results that will be no more than 10 MB (the same size as our logging buffer which is held in memory).